pywhy_graphs/algorithms/generic.py

```diff
@@ -333,3 +334,210 @@ def _single_shortest_path_early_stop(G, firstlevel, paths, cutoff, join, valid_p
+        The current node.
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path which is empty if the bool is false.
+    """
+    path_exists = False
+    path = []
+    visited.add(cur_node)
+    children = G.neighbors(cur_node)
+    print(cur_node)
+    if cur_node is node_y:
+        return (True, [node_y])
+    for elem in children:
+        if elem in visited:
+            continue
+        else:
+            if (
+                _is_collider(G, prev_node, cur_node, elem)
+                and (cur_node not in check_ancestors)
+                and (cur_node not in S)
+                and (cur_node is not node_y)
+            ):
+                continue
+            elif (
+                not _is_collider(G, prev_node, cur_node, elem)
+                and (cur_node not in L)
+                and (cur_node is not node_y)
+            ):
+                continue
+            path_exists, temp_path = _recursive_path(
+                G, node_x, node_y, L, S, visited, check_ancestors, elem, cur_node
+            )
+            if path_exists:
+                path.append(cur_node)
+                path.extend(temp_path)
+                break
+    return (path_exists, path)
+
+
+def inducing_path(G, node_x, node_y, L=None, S=None):
+    """Checks if an inducing path exists between two nodes as defined in :footcite:`Zhang2008`.
+
+    Parameters
+    ----------
+    G : Graph
+        The graph.
+    node_x : node
+        The source node.
+    node_y : node
+        The destination node.
+    L : set
+        Nodes that are ignored on the path. Defaults to an empty set. See Notes for details.
+    S:  set
+        Nodes that are always conditioned on. Defaults to an empty set. See Notes for details.
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path if the bool is true, an empty list otherwise.
+
+    Notes
+    -----
+    An inducing path intuitively is a path between two non-adjacent nodes that
+    cannot be d-separated. Therefore, the path is always "active" regardless of
+    what variables we condition on. L contains all the non-colliders, these nodes
+    are ignored along the path. S contains all the colliders, these nodes and their
+    ancestors are conditioned upon.
+    """
+    if L is None:
+        L = set()
+
+    if S is None:
+        S = set()
+
+    nodes = set(G.nodes)
+
+    if node_x not in nodes or node_y not in nodes:
+        raise ValueError("The provided nodes are not in the graph.")
+
+    if node_x == node_y:
+        raise ValueError("The start and destination nodes are the same.")
+
+    path = []  # this will contain the path.
+
+    xanc = _directed_sub_graph_parents(G, node_x)
+    yanc = _directed_sub_graph_parents(G, node_y)
+
+    xyancestors = xanc.union(yanc)
+
+    check_ancestors = set()
+
+    for elem in S:
+        check_ancestors = check_ancestors.union(_directed_sub_graph_parents(G, elem))
```

Comment:
This only gets parents, so:

```
X -> Z <- Y, A <- B <- Z

with 
S = {'A'}

would fail since X has an inducing path to Y with respect to L={} and S={'A'}
```

Reply:
Oh wait, so an ancestor of any node is a node that has a directed path to the said node?

Reply:
Conditioning on a descendant of a collider opens the collider